iceberg/src/main/scala/com/nvidia/spark/rapids/iceberg/parquet/GpuParquetReaderPostProcessor.scala

```diff
@@ -0,0 +1,169 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids.iceberg.parquet
+
+import java.util.{Map => JMap}
+
+import scala.collection.mutable.ArrayBuffer
+
+import com.nvidia.spark.rapids.{CastOptions, GpuCast, GpuColumnVector, GpuScalar}
+import com.nvidia.spark.rapids.Arm.withResource
+import com.nvidia.spark.rapids.iceberg.parquet.GpuParquetReaderPostProcessor.{doUpCastIfNeeded, HandlerResult}
+import java.util
+import org.apache.iceberg.{MetadataColumns, Schema}
+import org.apache.iceberg.shaded.org.apache.parquet.schema.MessageType
+import org.apache.iceberg.spark.SparkSchemaUtil
+import org.apache.iceberg.types.{Type, Types, TypeUtil}
+import org.apache.iceberg.types.Types.NestedField
+
+import org.apache.spark.sql.types.DataType
+import org.apache.spark.sql.vectorized.{ColumnarBatch, ColumnVector}
+
+/** Processes columnar batch after reading from parquet file.
+ *
+ * Apache iceberg uses a lazy approach to deal with schema evolution, e.g. when you
+ * add/remove/rename a column, it's essentially just metadata operation, without touching data
+ * files. So after reading from parquet, we need to deal with missing column, type promotion,
+ * etc. And these are all handled in [[GpuParquetReaderPostProcessor]].
+ *
+ * For details of schema evolution, please refer to
+ * [[https://iceberg.apache.org/spec/#schema-evolution iceberg spec]].
+ *
+ * @param fileReadSchema Schema passed to actual parquet reader.
+ * @param idToConstant   Constant fields.
+ * @param expectedSchema Iceberg schema required by reader.
+ */
+class GpuParquetReaderPostProcessor(
+    private[parquet] val fileReadSchema: MessageType,
+    private[parquet] val idToConstant: JMap[Integer, _],
+    private[parquet] val expectedSchema: Schema
+) {
+  require(fileReadSchema != null, "fileReadSchema cannot be null")
+  require(idToConstant != null, "idToConstant cannot be null")
+  require(expectedSchema != null, "expectedSchema cannot be null")
+
+  /**
+   * Process columnar batch to match expected schema.
+   *
+   * @param originalBatch Columnar batch read from parquet. Note that <b>the caller</b> should take
+   *                      care of managing ownership of it.
+   * @return Processed columnar batch.
+   */
+  def process(originalBatch: ColumnarBatch): ColumnarBatch = {
+    require(originalBatch != null, "Columnar batch can't be null")
+    require(fileReadSchema.getFieldCount == originalBatch.numCols(),
+      s"File read schema field count ${fileReadSchema.getFieldCount} doesn't match expected " +
+        s"columnar batch columns ${originalBatch.numCols()}")
+
+    withResource(new ColumnarBatchHandler(this, originalBatch)) { handler =>
```

Comment:
We already have a SchemaUtils class that handles type conversion for other readers.

https://github.com/NVIDIA/spark-rapids/blob/branch-25.06/sql-plugin/src/main/scala/com/nvidia/spark/rapids/SchemaUtils.scala

Is there a reason we need a TypeUtil to effectively do the same job?

Reply:
I think all of those are things that our Parquet code should probably end up supporting. Could you please file a follow on issue for us to figure out how to combine the two implementations?